include/yaclib/async/future.hpp

```diff
@@ -0,0 +1,315 @@
+#pragma once
+
+#include <yaclib/async/core.hpp>
+#include <yaclib/executor/executor.hpp>
+#include <yaclib/util/type_traits.hpp>
+
+#include <cassert>
+#include <condition_variable>
+#include <iostream>
+#include <mutex>
+#include <thread>
+
+namespace yaclib::async {
+
+template <typename T>
+class Future;
+
+namespace detail {
+
+template <typename Functor, typename Arg>
+struct Invoke {
+  using type = std::invoke_result_t<Functor, Arg>;
+};
+
+template <typename Functor>
+struct Invoke<Functor, void> {
+  using type = std::invoke_result_t<Functor>;
+};
+
+template <typename Functor, typename Arg>
+using InvokeT = typename Invoke<Functor, Arg>::type;
+
+template <typename Functor, typename Arg>
+struct IsInvocable {
+  constexpr static bool value = std::is_invocable_v<Functor, Arg>;
+};
+
+template <typename Functor>
+struct IsInvocable<Functor, void> {
+  constexpr static bool value = std::is_invocable_v<Functor>;
+};
+
+template <typename Functor, typename Arg>
+inline constexpr bool IsInvocableV = IsInvocable<Functor, Arg>::value;
+
+template <typename T>
+inline constexpr bool IsFutureV = std::integral_constant<bool, util::IsInstantiationOf<Future, T>::value>::value;
+
+template <typename U>
+struct FutureValue {
+  using type = U;
+};
+
+template <typename U>
+struct FutureValue<Future<U>> {
+  using type = U;
+};
+
+template <typename T>
+using FutureValueT = typename FutureValue<T>::type;
+
+}  // namespace detail
+
+template <typename T>
+class Future final {
+  template <typename U>
```

Comment:
- [ ] Make static_assert public as part of API

- [ ] Remove private section before the public and move a friend to the end